battenberg_gitlab/apply.py

```diff
@@ -0,0 +1,101 @@
+    templates = json.load(f)
+
+def construct_template(project_name: str) -> dict:
+    template = {
+        **templates['defaults'],
+        'project_name': project_name,
+        'module_name': project_name.replace('-', '_'),
+        **templates['projects'][project_name]
+    }
+    logging.info(f'Constructing template for {project_name}')
+    # logging.info(template)
+    return template
+
+
+def add_template(project_name: str, local_path: str, repo: Repository, merge_target: str):
+    # TODO Remove these once we've we've backfilled .cookiecutter.json
+    cookiecutter_json_path = os.path.join(local_path, '.cookiecutter.json')
+    logger.info(f'Creating .cookiecutter.json file at {cookiecutter_json_path}')
+    template = construct_template(project_name)
+    with open(cookiecutter_json_path, 'w') as f:
+        json.dump(template, f, sort_keys=True, indent=4)
+
+    # Create a new merge_target branch.
+    merge_target_ref = f'refs/heads/{merge_target}'
+    repo.branches.local.create(merge_target, repo.get(repo.head.target))
+    repo.checkout(merge_target_ref)
+
+    # Actually write the .cookiecutter.json to a new commit on the merge_target branch.
+    repo.index.add_all()
+    repo.index.write()
+    tree = repo.index.write_tree()
+    repo.create_commit(
+        'HEAD',
+        repo.default_signature,
+        repo.default_signature,
+        'Add .cookiecutter.json',
+        tree,
+        [repo.head.target]
+    )
+    repo.state_cleanup()
+    repo.checkout('HEAD')
+
+
+def apply(gl: Gitlab,
+          group_ids: List[int] = None,
+          project_filter: str = None,
+          workspace_path: str = None,
+          checkout: str = None):
+    projects = get_projects(gl, group_ids, project_filter)
+    if not projects:
+        raise Exception('No projects found')
+
+    # Create a working directory where we can clone all the projects we've found into.
+    workspace_path = ensure_workspace(workspace_path)
+    logger.info(f'Created workspace at {workspace_path}')
+
+    merge_conflicts = []
+    for project in projects:
+        local_path = os.path.join(workspace_path, project.name)
+        logger.info(f'Cloning {project.ssh_url_to_repo} to {local_path}')
+        repo = clone_or_discover_repo(project.ssh_url_to_repo, local_path)
+        
+        # TODO Remove add_template once this merge is over.
+        merge_target = f'archetype-update-{checkout}'
+        try:
+            add_template(project.name, local_path, repo, merge_target)
+        except KeyError as e:
+            logger.error(f'Skipping {project.name}')
+            logger.error(e)
+            continue
+
+        battenberg = Battenberg(repo)
+        try:
+            battenberg.upgrade(checkout=checkout, no_input=True, merge_target=merge_target)
+        except MergeConflictException as e:
+            logger.error(f'Merge conflicts found for {project.name} in {local_path}')
```

Comment:
Aren't merge conflicts expected? A log level of `error` seems inappropriate for something that should routinely happen. Maybe an `info` level makes more sense.

Reply:
Not necessarily. In theory there could be updates we push out, ie. just updating pins for example where if the existing line isn't changed it should be possible to merge, create a new commit, push the `merge-target` branch and create a PR in Gitlab using `battenberg-gitlab` :)

That should in theory be possible for v1.1.1 for example.

Reply:
I realize that we won't get merge conflicts in all cases, but they're a normal part of the flow. I wouldn't consider them errors.

Reply:
I'm not sure how this was resolved? Even in demos, Steven observed the merge conflicts logged as errors and came to the same conclusion.

Reply:
A warn level might be a good compromise here. I imagine there are some battenberg releases that we expect no merge conflicts (think: base image upgrade), so a large number of warnings would be concerning if I were wearing the release hat in that scenario.

Reply:
Cool switched to use a warning. I think this interface could definitely use some updates as it needs a little bit of simplification as it's currently pretty verbose. I'll see if I can clean it up in my part 2 PR that'll introduce a `click` CLI.